### DISCUSSION TITLE
What is the difference between homography_warp and warp_perspective?

### DISCUSSION THREAD

**Discussion by darkAlert** [kornia/kornia#849] [Questions] [answered]
I've trained a model that predicts homography (3x3 matrix). I used the homography_warp function to warp the input image with predicted homography. I then used this warped image to calculate the loss. Finaly, the model's learnt to predict homogrphy.

However, I thought I could use this predicted homography in the cv.warpPerspective (or in kornia.warp_perspective) but I was wrong, I guess...

Can someone explain to me what is the difference between the transformation matrix using in kornia.homography_warp and kornia.warp_perspective (cv.warpPerspective), please?

**Comment by edgarriba** [comment]
the first assumes a normalized homography and the second it's assumed to map in pixel space

**Comment by darkAlert** [comment]
@edgarriba 
Sorry, I don't understand. The first and second take an image and a 3x3 matrix as input and output the image as the result. Is there a way to convert the input matrix from the first to the second?

**Comment by darkAlert** [comment]
Or is there a way to project some points with this input matrix that I use in homography_warp?

**Comment by edgarriba** [comment]
@darkAlert  you can use this matrix to convert https://kornia.readthedocs.io/en/latest/geometry.warp.html?highlight=normal_transform_pixel#kornia.geometry.warp.normal_transform_pixel

**Reply by darkAlert** [reply]
Sorry, but I still don't understand how to use kornia.geometry.warp.normal_transform_pixel in my case

**Comment by darkAlert** [comment, marked as answer]
finally, I've found a solution:
1) normalize the input points in the range [-1,1] (norm_pts=(pts-0.5)*2)
2) take the inverse of the homography matrix  
3) apply transformation to the normalized points  (via kornia.geometry.linalg.transform_points or cv2.perspectiveTransform or by hands)
4) apply inverse normalization to the transformed points (pts = trans_pts/2+0.5)

**Reply by edgarriba** [reply]
Another solution is the following:
```python
dst_H_src: torch.Tensor  # is your homography that maps points from a source to destination view
norm_T_pix =  kornia.geometry.warp.normal_transform_pixel(height, width)  # transform that normalises the points from pixel to normalised space
H_new = norm_T_pix.inverse() H @ norm_T_pix
```
and then use `H_new`
